File: 300.最长上升子序列/lengthOfLIS.go
package problem0300

// 数组含义	dp[i]的定义是【以nums[i]结尾的最长上升子序列的长度】, 搞清楚这点很关键, [5,9,3] => dp[2]值为1, 代表的是以3结尾的上升子序列长度为1
// 初始值	dp[i] = 1, 每个元素自身构成
// 转移方程	dp[i] = max(dp[i], dp[j]+1)
// 答案		max(dp)
func lengthOfLIS(nums []int) int {
	dp := make([]int, len(nums))
	for i := 0; i < len(nums); i++ {
		dp[i] = 1
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
	}
	result := 0
	for i := 0; i < len(nums); i++ {
		result = max(result, dp[i])
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 300.最长上升子序列/lengthOfLIS.js
// LIS
// 通常会想：dp[i][j] 代表 nums[i:j] 的 LIS的长度，但是这样推导不出来
// 正确定义：dp[i] 代表以 nums[i] 结尾的 LIS的长度
// 显然有初始值: dp[i] = 1
// 状态转移方程：dp[i] = max(dp[j]) + 1 (0 <= j < i && nums[j] < nums[i])

// 时间复杂度 O(n^2)
// 空间复杂度 O(n)
function lengthOfLIS(nums) {
    const l = nums.length
    const dp = new Array(l).fill(1)
    let result = 0

    for (let i = 0; i < l; i++) {
        let max = 0
        for (let j = 0; j < i; j++) {
            if (nums[j] < nums[i]) {
                max = Math.max(max, dp[j])
            }
        }
        dp[i] = max + 1
        result = Math.max(result, dp[i])
    }

    return result
}


// 二分
// 时间复杂度O(nlogn)
// 空间复杂度O(n)
function lengthOfLIS(nums) {
    const n = nums.length
    if (n <= 1) {
        return n
    }
    const d = [nums[0]]

    for (let i = 1; i < n; i++) {
        if (nums[i] > d[d.length - 1]) {
            d.push(nums[i])
            continue
        }
        let l = 0
        let r = d.length - 1
        while (l < r) {
            const mid = (l + r) >> 1
            if (d[mid] < nums[i]) {
                l = mid + 1
            } else {
                r = mid
            }
        }
        d[l] = nums[i]
    }
    return d.length
}

// 求LIS的一个解
function oneOfLIS(nums) {
    const n = nums.length
    if (n <= 1) {
        return nums
    }
    const prev = nums.slice()
    const result = [0] // 记录索引

    for (let i = 1; i < n; i++) {
        if (nums[result[result.length - 1]] < nums[i]) {
            prev[i] = result[result.length - 1]
            result.push(i)
            continue
        }
        let left = 0
        let right = result.length - 1
        while (left < right) {
            const mid = (left + right) >> 1
            if (nums[result[mid]] < nums[i]) {
                left = mid + 1
            } else {
                right = mid
            }
        }
        if (left > 0) {
            prev[i] = result[left - 1]
        }
        result[left] = i
    }
    let len = result.length
    let val = result[len - 1]
    while (len-- > 0) {
        result[len] = val
        val = prev[val]
    }
    return result
}
